agent.go

```diff
@@ -1172,50 +1181,155 @@ func (c *container) removeContainer(id string) error {
 	return unmountContainerRootFs(id, c.config.Rootfs)
 }
 
-func removeContainerCb(pod *pod, data []byte) error {
+func (c *container) processList(id string, format string, args []string) ([]byte, error) {
+	pids, err := c.container.Processes()
+	if err != nil {
+		return nil, err
+	}
+
+	switch format {
+	case "table":
+	case "json":
+		return json.Marshal(pids)
+	default:
+		return nil, fmt.Errorf("invalid format option")
+	}
+
+	psArgs := args
+	if len(psArgs) == 0 {
+		psArgs = []string{"-ef"}
+	}
+
+	cmd := exec.Command("ps", psArgs...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", err, output)
+	}
+
+	lines := strings.Split(string(output), "\n")
+
+	// looking for PID field in ps title
+	pidIndex := -1
+	fields := strings.Fields(lines[0])
+	for i, f := range fields {
+		if f == "PID" {
+			pidIndex = i
+			break
+		}
+	}
+
+	// PID field not found
+	if pidIndex == -1 {
+		return nil, fmt.Errorf("failed to find PID field in ps output")
+	}
+
+	// append title
+	var result bytes.Buffer
+
+	result.WriteString(lines[0] + "\n")
+
+	for _, line := range lines[1:] {
+		if len(line) == 0 {
+			continue
+		}
+		fields := strings.Fields(line)
+		if pidIndex >= len(fields) {
+			return nil, fmt.Errorf("missing PID field: %s", line)
+		}
+
+		p, err := strconv.Atoi(fields[pidIndex])
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert pid to int: %s", fields[pidIndex])
+		}
+
+		// appends pid line
+		for _, pid := range pids {
+			if pid == p {
+				result.WriteString(line + "\n")
+				break
+			}
+		}
+	}
+
```

Comment:
I get that this whole function is about duplicating what `runc` does here https://github.com/opencontainers/runc/blob/master/ps.go#L41-L95, but I think we should add an added value, that is gather all those data inside a structure dedicated to this. This structure would be then marshalled so that it would become a JSON byte array that we can send to the runtime(virtcontainers basically).

Reply:
do you mean to put all output inside a struct?

```
type psOutput struct {
    output byte[]
}
```

I wonder if this is really needed

Reply:
@devimc no, I meant having something nicer than only an array of byte already in the expected format. Basically, I was thinking that the agent could parse the output of the `ps` command to put it into a structure like:
```go
type processInfo struct {
        Name string
        PID     string
        ...
}

type processListInfo []processInfo
```
And this would be marshalled into a json byte array through the response. Then we would have virtcontainers relying on this specific structure to unmarshal the byte array. And then, this structure would be passed back to the runtime. The runtime wrapper being the only one doing the formatting.

I am saying all of this because if we think about virtcontainers as a generic library, we want to return a process list as a standardized structure that any consumer of virtcontainers could format the way it wants.

Reply:
@sboeuf  the problem with the structure that you propose is that we don't know the options user will provide

Reply:
That's why we need to have every potential output covered by this structure. The output will depend on the options, I agree, but eventually we would be sending the same structure.

Reply:
the list of possible formats/columns is huge, I think  that is not a good idea. I wouldn't like to update/maintain this structure each time a new column/format is added/removed in the ```ps``` command

Reply:
@devimc Yes I understand...
@sameo can we get your input on this please ?

Reply:
@devimc I'd agree with @sboeuf here. The agent should send data without assuming anything about the end user expected format. The runtime should be the one doing the formatting/filtering, not the agent.

Reply:
@sameo great! , so can we merge?

Reply:
@devimc I don't think you're marshalling an actual structure that the runtime/virtcontainers could unmarshal and format filter, right ?

Reply:
```runc``` doesn't format/filter the output because docker does that [1], I consider we should have a similar implementation and leave to docker decide the output format.

[1] - https://github.com/moby/moby/blob/402540708c9a0c35dc0b279a0f330455633537b8/daemon/top_unix.go#L63

Reply:
All right, let's get this one in.